### ISSUE TITLE
Error updating ScaleObject: the object has been modified

### ISSUE THREAD

**Issue by aminebizid** [kedacore/keda#158]
I get this error when trying to scale active ScaleObject

```
time="2019-05-07T20:02:31Z" level=error msg="Error updating scaledObject (hpa/azure-servicebus-queue-scaledobject) status: Operation cannot be fulfilled on scaledobjects.keda.k8s.io \"azure-servicebus-queue-scaledobject\": the object has been modified; please apply your changes to the latest version and try again"
```

The error is thrown here:
scale_handler.go:269
```go
func (h *ScaleHandler) updateScaledObject(scaledObject *keda_v1alpha1.ScaledObject) error {
	_, err := h.kedaClient.KedaV1alpha1().ScaledObjects(scaledObject.GetNamespace()).Update(scaledObject)
	if err != nil {
		log.Errorf("Error updating scaledObject (%s/%s) status: %s", scaledObject.GetNamespace(), scaledObject.GetName(), err.Error())
	}
	return err
}
```

**Comment by yaron2** [comment]
I think I know why this might be happening.
Did if actually affect scaling in any way, though?

**Comment by aminebizid** [comment]
It scales only once from 0 to 1

**Comment by aminebizid** [comment]
My cluster scaling configs

horizontal-pod-autoscaler-sync-period | 30s
-- | --
horizontal-pod-autoscaler-downscale-delay | 5m0s 
horizontal-pod-autoscaler-upscale-delay | 3m0s

**Comment by yaron2** [comment]
Is that error showing up repeatedly in the logs?

**Comment by aminebizid** [comment]
yes it's showing repeatedly  and I get <unknown> in the HPA
```bash
k -n hpa get hpa
NAME                REFERENCE             TARGETS              MINPODS   MAXPODS   REPLICAS   AGE
keda-hpa-consumer   Deployment/consumer   <unknown>/30 (avg)   1         10        1          7h
```

**Comment by yaron2** [comment]
Can you try to delete the keda controller pod running under the keda namespace, wait for it to come back up and see if the error keeps popping up?

**Comment by aminebizid** [comment]
Same behaviour.
How do you do to debug the container running in the cluster?

**Comment by yaron2** [comment]
You can try and debug locally by running the setup on minikube.

Can you delete and recreate the scaledobject?
Also, what trigger are you using?

**Comment by aminebizid** [comment]
Already done hundreds time 

my templates:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: consumer
  labels:
    app: consumer
spec:
  selector:
    matchLabels:
      app: consumer
  template:
    metadata:
      labels:
        app: consumer
    spec:
      containers:
      - name: consumer
        image: aminrbizid/servicebus-topic:1.0.0
        args: ["testk8s", "testk8ssub"]
        env:
          - name: SERVICEBUS_CONNECTION_STRING
            valueFrom:
              secretKeyRef:
                name: servicebuskey
                key: sb-connection-string
---
apiVersion: keda.k8s.io/v1alpha1
kind: ScaledObject
metadata:
  name: azure-servicebus-queue-scaledobject
  namespace: hpa
  label:
    deploymentName: consumer
spec:
  scaleTargetRef:
    deploymentName: consumer
  pollingInterval: 5
  minReplicaCount: 1
  maxReplicaCount: 10 # Optional. Default: 100
  triggers:
  - type: azure-servicebus
    metadata:
      # Required: queueName OR topicName and subscriptionName
      # queueName: functions-sbqueue
      # or
      topicName: testk8s
      subscriptionName: testk8ssub
      # Required
      connection: SERVICEBUS_CONNECTION_STRING
      # Optional
      queueLength: "30" # default 5
```

**Comment by aminebizid** [comment]
after delete keda pod:

```

time="2019-05-07T20:52:03Z" level=info msg="ActiveMessageCount :771"
time="2019-05-07T20:52:03Z" level=error msg="Error updating scaledObject (hpa/azure-servicebus-queue-scaledobject) status: Operation cannot be fulfilled on scaledobjects.keda.k8s.io \"azure-servicebus-queue-scaledobject\": the object has been modified; please apply your changes to the latest version and try again"
time="2019-05-07T20:52:08Z" level=info msg="Entring GetAzureServiceBusLength"
time="2019-05-07T20:52:08Z" level=info msg="Entring GetSubscriptionEntityFromNamespace"
time="2019-05-07T20:52:08Z" level=info msg="ActiveMessageCount :1083"
time="2019-05-07T20:52:08Z" level=error msg="Error updating scaledObject (hpa/azure-servicebus-queue-scaledobject) status: Operation cannot be fulfilled on scaledobjects.keda.k8s.io \"azure-servicebus-queue-scaledobject\": the object has been modified; please apply your changes to the latest version and try again"
```

**Comment by Aarthisk** [comment]
Can you share the output of kubectl describe hpa keda-hpa-consumer?

**Comment by aminebizid** [comment]
```bash
$ k -n hpa describe hpa

Name:                                    keda-hpa-consumer
Namespace:                               hpa
Labels:                                  <none>
Annotations:                             <none>
CreationTimestamp:                       Wed, 08 May 2019 16:07:33 +0200
Reference:                               Deployment/consumer
Metrics:                                 ( current / target )
  "queueLength" (target average value):  <unknown> / 30
Min replicas:                            1
Max replicas:                            10
Conditions:
  Type           Status  Reason                   Message
  ----           ------  ------                   -------
  AbleToScale    True    SucceededGetScale        the HPA controller was able to get the target's current scale
  ScalingActive  False   FailedGetExternalMetric  the HPA was unable to compute the replica count: unable to get external metric hpa/queueLength/&LabelSelector{MatchLabels:map[string]string{deploymentName: consumer,},MatchExpressions:[],}: unable to fetch metrics from external metrics API: Exactly one scaled object should match label deploymentName=consumer
Events:
  Type     Reason                        Age                From                       Message
  ----     ------                        ----               ----                       -------
  Warning  FailedGetExternalMetric       6s (x10 over 16h)  horizontal-pod-autoscaler  unable to get external metric hpa/queueLength/&LabelSelector{MatchLabels:map[string]string{deploymentName: consumer,},MatchExpressions:[],}: unable to fetch metrics from external metrics API: Exactly one scaled object should match label deploymentName=consumer
  Warning  FailedComputeMetricsReplicas  6s (x10 over 16h)  horizontal-pod-autoscaler  failed to get queueLength external metric: unable to get external metric hpa/queueLength/&LabelSelector{MatchLabels:map[string]string{deploymentName: consumer,},MatchExpressions:[],}: unable to fetch metrics from external metrics API: Exactly one scaled object should match label deploymentName=consumer

```

**Comment by Aarthisk** [comment]
Can you share the scaled objects deployed to minikube (if there are more than 1 with the same deployment name)? When the HPA queries the metrics adapter for the metrics, the metrics adapter queries for the scaled objects matching the label deploymentName=*** If exactly one scaled object does not match (so 0 or 1+) then this error gets returned. I have seen this error when the scaled object deployed likely does not have the deploymentName label. (your above scaled object looks OK but is this the one as minikube as well?)

**Comment by jeffhollan** [comment]
I think this may be a bit of a double-bug.  I get this error after the first activation where it fails to update scaledObject - repros consistently just following the tutorial for the RabbitMQ and Go sample.  However my scaling does work and the error shows repeatedly in the logs whenever a deployment has been activated.  The logs stop repeating after it scales down to 0.

Deleted and recreated entire setup and it happened again.

However, I think @aminebizid is likely hitting this bug but ALSO has another issue that may be impacting his HPA.  It reads in that his `ScaledObject` may not have all the right labels set?  @aminebizid  it'd be good if you could provide a sample of the `deployment` and the `scaledObject` definitions you are applying.

I think there is still a bug though in this error getting thrown every 5 seconds though which appears to be "harmless"

**Comment by aminebizid** [comment]
@jeffhollan you are right, I get 2 kind of errors and the second one is fatal:

```
E0510 09:31:57.022163       1 status.go:64] apiserver received an error that is not an metav1.Status: &errors.errorString{s:"Exactly one scaled object should match label deploymentName=consumer"}
```

```
time="2019-05-10T09:31:59Z" level=error msg="Error updating scaledObject (hpa/azure-servicebus-queue-scaledobject) status: Operation cannot be fulfilled on scaledobjects.keda.k8s.io \"azure-servicebus-queue-scaledobject\": the object has been modified; please apply your changes to the latest version and try again" 
```

Here are my used templates:
If you want, I can send you my bus topic  connection string in a private message.


### consumer
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: consumer
  namespace: hpa
  labels:
    app: consumer
spec:
  selector:
    matchLabels:
      app: consumer
  template:
    metadata:
      labels:
        app: consumer
    spec:
      containers:
      - name: consumer
        image: aminrbizid/servicebus-topic:1.0.0
        args: ["testk8s", "testk8ssub"]
        env:
          - name: SERVICEBUS_CONNECTION_STRING
            valueFrom:
              secretKeyRef:
                name: servicebuskey
                key: sb-connection-string
    
```

### scaledObject

```yaml
apiVersion: keda.k8s.io/v1alpha1
kind: ScaledObject
metadata:
  name: azure-servicebus-queue-scaledobject
  namespace: hpa
  label:
    deploymentName: consumer
spec:
  scaleTargetRef:
    deploymentName: consumer
  pollingInterval: 5
  minReplicaCount: 1
  maxReplicaCount: 10 # Optional. Default: 100
  triggers:
  - type: azure-servicebus
    metadata:
      # Required: queueName OR topicName and subscriptionName
      # queueName: functions-sbqueue
      # or
      topicName: testk8s
      subscriptionName: testk8ssub
      # Required
      connection: SERVICEBUS_CONNECTION_STRING
      # Optional
      queueLength: "30" # default 5
```

I create the secret in hpa namespace like follow:

```bash
k -n hpa create secret generic servicebuskey --from-literal=sb-connection-string=$SERVICEBUS_CONNECTION_STRING
```

```bash
k -n hpa describe scaledObject azure-servicebus-queue-scaledobject
```
```yaml
Name:         azure-servicebus-queue-scaledobject
Namespace:    hpa
Labels:       <none>
Annotations:  kubectl.kubernetes.io/last-applied-configuration={"apiVersion":"keda.k8s.io/v1alpha1","kind":"ScaledObject","metadata":{"annotations":{},"label":{"deploymentName":"consumer"},"name":"azure-servicebus-...
API Version:  keda.k8s.io/v1alpha1
Kind:         ScaledObject
Metadata:
  Creation Timestamp:  2019-05-10T09:13:10Z
  Generation:          1
  Resource Version:    7484
  Self Link:           /apis/keda.k8s.io/v1alpha1/namespaces/hpa/scaledobjects/azure-servicebus-queue-scaledobject
  UID:                 d4c97391-7303-11e9-a5c7-0800275810f5
Spec:
  Max Replica Count:  10
  Min Replica Count:  1
  Polling Interval:   5
  Scale Target Ref:
    Deployment Name:  consumer
  Triggers:
    Metadata:
      Connection:         SERVICEBUS_CONNECTION_STRING
      Queue Length:       30
      Subscription Name:  testk8ssub
      Topic Name:         testk8s
    Type:                 azure-servicebus
Events:                   <none>
```
```bash
k -n hpa get hpa
NAME                REFERENCE             TARGETS              MINPODS   MAXPODS   REPLICAS   AGE
keda-hpa-consumer   Deployment/consumer   <unknown>/30 (avg)   1         10        0          10s
```
```bash
k -n hpa describe hpa keda-hpa-consumer
```
```yaml
Name:                                    keda-hpa-consumer
Namespace:                               hpa
Labels:                                  <none>
Annotations:                             <none>
CreationTimestamp:                       Fri, 10 May 2019 11:13:10 +0200
Reference:                               Deployment/consumer
Metrics:                                 ( current / target )
  "queueLength" (target average value):  <unknown> / 30
Min replicas:                            1
Max replicas:                            10
Conditions:
  Type           Status  Reason                   Message
  ----           ------  ------                   -------
  AbleToScale    True    SucceededGetScale        the HPA controller was able to get the target's current scale
  ScalingActive  False   FailedGetExternalMetric  the HPA was unable to compute the replica count: unable to get external metric hpa/queueLength/&LabelSelector{MatchLabels:map[string]string{deploymentName: consumer,},MatchExpressions:[],}: unable to fetch metrics from external metrics API: Exactly one scaled object should match label deploymentName=consumer
Events:
  Type     Reason                        Age               From                       Message
  ----     ------                        ----              ----                       -------
  Warning  FailedComputeMetricsReplicas  4m (x12 over 7m)  horizontal-pod-autoscaler  failed to get queueLength external metric: unable to get external metric hpa/queueLength/&LabelSelector{MatchLabels:map[string]string{deploymentName: consumer,},MatchExpressions:[],}: unable to fetch metrics from external metrics API: Exactly one scaled object should match label deploymentName=consumer
  Warning  FailedGetExternalMetric       2m (x21 over 7m)  horizontal-pod-autoscaler  unable to get external metric hpa/queueLength/&LabelSelector{MatchLabels:map[string]string{deploymentName: consumer,},MatchExpressions:[],}: unable to fetch metrics from external metrics API: Exactly one scaled object should match label deploymentName=consumer
```

```bash
kubectl get --raw /apis/keda.k8s.io/v1alpha1/namespaces/hpa/scaledobjects/azure-servicebus-queue-scaledobject | jq
```
```json
{
  "apiVersion": "keda.k8s.io/v1alpha1",
  "kind": "ScaledObject",
  "metadata": {
    "annotations": {
      "kubectl.kubernetes.io/last-applied-configuration": "{\"apiVersion\":\"keda.k8s.io/v1alpha1\",\"kind\":\"ScaledObject\",\"metadata\":{\"annotations\":{},\"label\":{\"deploymentName\":\"consumer\"},\"name\":\"azure-servicebus-queue-scaledobject\",\"namespace\":\"hpa\"},\"spec\":{\"maxReplicaCount\":10,\"minReplicaCount\":1,\"pollingInterval\":5,\"scaleTargetRef\":{\"deploymentName\":\"consumer\"},\"triggers\":[{\"metadata\":{\"connection\":\"SERVICEBUS_CONNECTION_STRING\",\"queueLength\":\"30\",\"subscriptionName\":\"testk8ssub\",\"topicName\":\"testk8s\"},\"type\":\"azure-servicebus\"}]}}\n"
    },
    "creationTimestamp": "2019-05-10T09:13:10Z",
    "generation": 1,
    "name": "azure-servicebus-queue-scaledobject",
    "namespace": "hpa",
    "resourceVersion": "7484",
    "selfLink": "/apis/keda.k8s.io/v1alpha1/namespaces/hpa/scaledobjects/azure-servicebus-queue-scaledobject",
    "uid": "d4c97391-7303-11e9-a5c7-0800275810f5"
  },
  "spec": {
    "maxReplicaCount": 10,
    "minReplicaCount": 1,
    "pollingInterval": 5,
    "scaleTargetRef": {
      "deploymentName": "consumer"
    },
    "triggers": [
      {
        "metadata": {
          "connection": "SERVICEBUS_CONNECTION_STRING",
          "queueLength": "30",
          "subscriptionName": "testk8ssub",
          "topicName": "testk8s"
        },
        "type": "azure-servicebus"
      }
    ]
  }
}
```

```bash
k -n hpa describe pod consumer-8ffc9799-6t6nh
```
```yaml
Name:               consumer-8ffc9799-6t6nh
Namespace:          hpa
Priority:           0
PriorityClassName:  <none>
Node:               minikube/10.0.2.15
Start Time:         Fri, 10 May 2019 11:12:19 +0200
Labels:             app=consumer
                    pod-template-hash=8ffc9799
Annotations:        <none>
Status:             Running
IP:                 172.17.0.6
Controlled By:      ReplicaSet/consumer-8ffc9799
Containers:
  consumer:
    Container ID:  docker://9acea622449e3445f8c6f2043f3bfade8b9be568dcf6764929df9f774a67500b
    Image:         aminrbizid/servicebus-topic:1.0.0
    Image ID:      docker-pullable://aminrbizid/servicebus-topic@sha256:7b6233615667785914afbe327bdaeb3b3ed847b22b98749072e47040acb85000
    Port:          <none>
    Host Port:     <none>
    Args:
      testk8s
      testk8ssub
    State:          Running
      Started:      Fri, 10 May 2019 11:12:23 +0200
    Ready:          True
    Restart Count:  0
    Environment:
      SERVICEBUS_CONNECTION_STRING:  <set to the key 'sb-connection-string' in secret 'servicebuskey'>  Optional: false
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-mxqnt (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             True
  ContainersReady   True
  PodScheduled      True
Volumes:
  default-token-mxqnt:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  default-token-mxqnt
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:
  Type    Reason     Age   From               Message
  ----    ------     ----  ----               -------
  Normal  Scheduled  16m   default-scheduler  Successfully assigned hpa/consumer-8ffc9799-6t6nh to minikube
  Normal  Pulling    16m   kubelet, minikube  pulling image "aminrbizid/servicebus-topic:1.0.0"
  Normal  Pulled     16m   kubelet, minikube  Successfully pulled image "aminrbizid/servicebus-topic:1.0.0"
  Normal  Created    16m   kubelet, minikube  Created container
  Normal  Started    16m   kubelet, minikube  Started container
```

**Comment by zach-dunton-sf** [comment]
The scaledobject yaml examples have a typo, .metadata.label should be .metadata.labels, I'm preparing a pull request for SQS right now and will include that also.

**Comment by jeffhollan** [comment]
Thanks @zach-dunton-sf - just opened a PR on the example one.  I'll also send you an invite to the repo as a member so I can assign you to the SQS issue. Great to hear you are working on that.

Will keep this issue open though as we still have the error in the title that does show up.  I also should note that I mentioned I see the error every 5 second. I also have a polling interval of 5 seconds setup for that example, so those may be related.

@Aarthisk Can I assign you to this bug?

**Comment by zach-dunton-sf** [comment]
I have the SQS bit done, but still have the issue from the title of this one also. I will post more if I find a root cause.

**Comment by GokGokalp** [comment]
This issue seems like done but I'm still getting this error while scaling: "level=error msg="Error updating scaledObject (default/test-consumer) status: Operation cannot be fulfilled on scaledobjects.keda.k8s.io \"test-consumer\": the object has been modified; please apply your changes to the latest version and try again""

Btw: scaling works correctly.

**Comment by yaron2** [comment]
Are you using the Helm chart? Its possible that the changes didnt get pushed to our ```:master``` image yet.

Best way to verify is to build your own KEDA controller and test with that one.

**Comment by GokGokalp** [comment]
> Are you using the Helm chart? Its possible that the changes didnt get pushed to our `:master` image yet.
> 
> Best way to verify is to build your own KEDA controller and test with that one.

Ah okey, I used the helm chart. It's okey now. Thanks!

**Comment by saralpa** [comment]
Hi,

I am facing the same issue. How you resolved it?

**Comment by rangarb885** [comment]
Hi , I have the same problem when i change the configuration on scaled object. I have KEDA 2.4 running on EKS 1.21.  I don't remember having this problem when i was using KEDA 2.3 on EKS 1.16. 

Does this error affect scaling?

**Comment by zroubalik** [comment]
@rangarb885 you are probably facing this issue, right? https://github.com/kedacore/keda/issues/2035
If so, comment there or try the patch if you can. To confirm that it is correct.

Thanks
